```{r}
setwd('/home/lien/data/skrypty/TFIIB/')
library("tidyverse")
source('functions.R')
```
##### define groups of genes
```{r}
group_sets=list('geneBody'=c(1:31))
down_thresh=3000
up_thresh=50000
Groups <- create_Groups(group_sets, down_thresh,up_thresh)

```

### load the metadata for Hi-C samples
```{r}
sample_info <- read_tsv('/home/lien/data/skrypty/TFIIB/samples_GEO_HiC.tsv',col_types='ffffff')
sample_info
```

### Upper panels - mean asymmetry ratios in TFIIB1-WT non-treated with confidence intervals
##### Pull the data from *sectors.tsv files
```{r}
sector_data <- lapply(1:12,function(i){
  read_tsv(paste0('/home/lien/data/TFIIB/GEO_upload/HiC/Sample',i,'_sectors.tsv'), col_types='cdddc')
}) %>% purrr::reduce(bind_rows) 

```
##### generate a single plot with mean asymmetry ratios for different groups, can be averaged across multiple samples. Here, TFIIB1-WT non-treated samples are chosen
```{r}

boot_plot <- bootANDplot(data=sector_data,Groups,merge_samples=c(1,9))

boot_plot[[1]]+
  scale_color_discrete(limits=names(Groups))+
     scale_x_discrete(limits=names(Groups))+
  theme(legend.position = "none",
        aspect.ratio=NULL)
```


### plot expresion for these groups of genes, based on 3'RNA-seq data
#####  readCounts_table.tsv can be downloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE199056
```{r}
nor <- read_tsv("readCounts_table.tsv")

sample_info_3RNAseq <- read_tsv('sample_info_3RNAseq.tsv', col_types ='cffdffff') %>%
  mutate(comparison=paste0(stress,'_',timepoint),
         genotype=factor(genotype, levels=c('Wt','R58C')))
sample_info_3RNAseq
```
##### calculate rpm, average of biological replicates. Take only data from TFIIB1-WT nontreated sample, remove genes that were not detected
```{r}
down_thresh=3000
up_thresh=50000
mean_expression <- calculate_mean_rpm(nor, filter_nuclear=T, filter_protCoding=T)
WT_NT_mean_rpm <- mean_expression[c('GeneID','Wt_S1_0')]%>%
   subset(Wt_S1_0>0)
```
##### divide genes into groups
```{r}
expression_data <- lapply(names(Groups),function(g){
WT_NT_mean_rpm %>%
  subset(GeneID %in% Groups[[g]])%>%
    subset(Wt_S1_0>1) %>%
  mutate(group=g) 
}) %>% purrr::reduce(bind_rows)
```



##### plot the result
```{r}
expression_data %>%
  ggviolin(x="group",y="Wt_S1_0",alpha=0.3,color='group',fill='group') +
  #scale_color_manual(values=kolory, aesthetics = c("color","fill"))+
   scale_y_log10(name='Expression level [rpm]',breaks=c(1,10,100,1000), labels=c('1','10','100','1000'))+
  #ylim(0,100)+
  #scale_x_discrete(limits=c("Excl_State 26","Excl_State 12", "All_PCG","Excl_State 25", "Excl_State 3" ))+
   geom_boxplot(aes(x=group,y=Wt_S1_0,color=group),width=0.1)+ 
  newtheme+
 theme(legend.position = "none",
        aspect.ratio=NULL,
              axis.text.y=element_text(angle = 90,hjust=0.5),
          axis.text.x=element_text(angle = 90,hjust=1),
          strip.text.y=element_text(angle = 90))
   
```

